File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
        return null;
        }
        // Use the 0-th list as a return list
        for (int i = 1; i < lists.length; ++i) {
            lists[0] = mergeTwoLists(lists[0], lists[i]);
        }

        return lists[0];
    }
    
    public ListNode solve(ListNode list1, ListNode list2){
        ListNode cur = list1;
        ListNode next = list1.next;
        if(next == null){
            cur.next = list2;
        }
        ListNode temp = list2;
        ListNode next2 = null;
        while(temp != null && next != null){
            
            if(cur.val <= temp.val && temp.val <= next.val){
                cur.next = temp;
                next2 = temp.next;
                temp.next = next;
                cur = temp;
                temp = next2;
            }
            else{
                cur = next;
                next = next.next;
                if(next == null){
                    cur.next = temp;
                    return list1;
                }
            }
            
        }
        return list1;
    }
    
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        if(list1 == null)
            return list2;
        if(list2 == null)
            return list1;
        
        if(list1.val <= list2.val){
            return solve(list1, list2);
        }
        
        else{
            return solve(list2, list1);
        } 
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* dummyNode = new ListNode(-1);
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> pq;
        for(int i = 0; i < lists.size(); i++){
            if(lists[i])
                pq.push({lists[i]->val, lists[i]});
        }
        ListNode* temp = dummyNode;
        while(!pq.empty()){
            int value = pq.top().first;
            ListNode* curr = pq.top().second;
            pq.pop();
            temp->next = curr;
            temp = curr;
            if(curr->next)
                pq.push({curr->next->val, curr->next});
        }
        return dummyNode->next;
    }
};